zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_7/DbDistributionMigrationState8dot7.java

```diff
@@ -0,0 +1,144 @@
+            transactionContext,
+            distributionKey,
+            new PersistedCommandDistribution());
+
+    partitionKey = new DbInt();
+    distributionPartitionKey = new DbCompositeKey<>(fkDistribution, partitionKey);
+    pendingDistributionColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.PENDING_DISTRIBUTION,
+            transactionContext,
+            distributionPartitionKey,
+            DbNil.INSTANCE);
+
+    retriableDistributionColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.RETRIABLE_DISTRIBUTION,
+            transactionContext,
+            distributionPartitionKey,
+            DbNil.INSTANCE);
+
+    queueId = new DbString();
+    queuePerPartitionKey = new DbCompositeKey<>(queueId, partitionKey);
+    queuedDistributionKey =
+        new DbCompositeKey<>(queueId, new DbCompositeKey<>(partitionKey, fkDistribution));
+    queuedCommandDistributionColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.QUEUED_DISTRIBUTION,
+            transactionContext,
+            queuedDistributionKey,
+            DbNil.INSTANCE);
+
+    continuationKey = new DbLong();
+    continuationByQueueKey = new DbCompositeKey<>(queueId, continuationKey);
+    continuationCommandColumnFamily =
+        zeebeDb.createColumnFamily(
+            ZbColumnFamilies.DISTRIBUTION_CONTINUATION,
+            transactionContext,
+            continuationByQueueKey,
+            new PersistedCommandDistribution());
+  }
+
+  public void migrateIdempotentCommandDistributions() {
+    queueId.wrapString(DistributionQueue.DEPLOYMENT.getQueueId());
+    final var isFirstInQueue = new AtomicBoolean(true);
+
+    pendingDistributionColumnFamily.forEach(
+        (compositeKey, nil) -> {
+          final var distributionKey = compositeKey.first().inner().getValue();
+          final var partitionId = compositeKey.second().getValue();
+          this.distributionKey.wrapLong(distributionKey);
+          partitionKey.wrapInt(partitionId);
+
+          final var persistedDistribution =
+              commandDistributionRecordColumnFamily.get(this.distributionKey);
+
+          final var valueType = persistedDistribution.getValueType();
+          final var isDeploymentOrDeletion =
+              valueType == ValueType.DEPLOYMENT || valueType == ValueType.RESOURCE_DELETION;
+
+          if (isDeploymentOrDeletion && persistedDistribution.getQueueId().isEmpty()) {
+            persistedDistribution.setQueueId(DistributionQueue.DEPLOYMENT.getQueueId());
+            commandDistributionRecordColumnFamily.update(
+                this.distributionKey, persistedDistribution);
+            queuedCommandDistributionColumnFamily.insert(queuedDistributionKey, DbNil.INSTANCE);
+
+            if (isFirstInQueue.get()) {
+              isFirstInQueue.set(false);
+            } else {
+              retriableDistributionColumnFamily.deleteExisting(distributionPartitionKey);
```

Comment:
The `retriableDistributionColumnFamily` is the CF we iterate over when we retry a distribution. Since we want these commands to be queued we only want to retry the first one. Only once this is ACK'd should we continue with the next command in the queue. Hence why we leave the first distribution we find in there, and remove the others from it.